File: tests/omp/teams1.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->

  !! Sequence.intro ~mark:"seq_ins" ~start:[tBefore; cVarDef "i"] ~stop:[tBefore; cReturn] ();
  
  !! Omp.teams ~clause:[Num_teams "2"] [cMark "seq_ins"];
  !! Omp.target ~clause:[Map_c (To, ["B[:N]";"C[:N]"]); Map_c (ToFrom, ["sum0";"sum1"])] [cMark "seq_ins"];
  
  !! Omp.parallel_for ~clause:[Reduction(Plus, ["sum0"])] [occIndex 0;cFor_c ""];
  !! Omp.parallel_for ~clause:[Reduction(Plus, ["sum1"])] [occIndex 1;cFor_c ""];

  !! Marks.clean [cMark "seq_ins"];
)

File: tests/omp/teams1.cpp
#include <stdlib.h>
#include <omp.h>
float dotprod(float B[], float C[], int N)
{
   float sum0 = 0.0;
   float sum1 = 0.0;
   int i;
   if (omp_get_num_teams() != 2)
      abort();
   if (omp_get_team_num() == 0)
   {
      for (i=0; i<N/2; i++)
         sum0 += B[i] * C[i];
   }
   else if (omp_get_team_num() == 1)
   {
      for (i=N/2; i<N; i++)
         sum1 += B[i] * C[i];
   }
   return sum0 + sum1;
}